### PR TITLE
Release - 2024-July

### PR DESCRIPTION
**Pull request by kopitek8**
### Previous release
- https://github.com/mapswipe/python-mapswipe-workers/pull/945

### Pull Requests
- https://github.com/mapswipe/python-mapswipe-workers/pull/955
- https://github.com/mapswipe/python-mapswipe-workers/pull/960

### Issues
- https://github.com/mapswipe/python-mapswipe-workers/issues/929
- https://github.com/mapswipe/python-mapswipe-workers/issues/947
- https://github.com/mapswipe/python-mapswipe-workers/issues/959

# Changelog
- Manager Dashboard
    - Add project name validation so that existing project names are not repeated #929 

### **NOTE: Do not delete `dev` branch**

### CODE REVIEW

File: manager-dashboard/app/views/NewProject/index.tsx [hunk 1]
```diff
@@ -271,9 +274,69 @@ function NewProject(props: Props) {
...
+                    setProjectSubmissionStatus(undefined);
+                    return;
+                }
+
+                const newProjectRef = await pushToDatabase(projectRef);
+                const newKey = newProjectRef.key;
+
+                if (!mountedRef.current) {
+                    return;
+                }
```
Comment (tnagorra) [inline manager-dashboard/app/views/NewProject/index.tsx:309]: We should do this after every await.

File: firebase/functions/src/index.ts [hunk 2]
```diff
@@ -343,3 +343,63 @@ exports.incProjectProgress = functions.database.ref('/v2/projects/{projectId}/re
...
+    try {
+        const projectRef = await admin.database().ref('v2/projects').once('value');
+        const data = projectRef.val();
+
+        if (data) {
+            const newProjectData: {[key: string]: string} = {};
+
+            Object.keys(data).forEach((id) => {
+                if (data[id]?.projectTopic) {
+                    const newProjectTopicKey = data[id].projectTopic?.toLowerCase() as string;
```
Comment (tnagorra) [inline firebase/functions/src/index.ts:357]: We might also strip out whitespace characters.

File: firebase/functions/src/index.ts [hunk 3]
```diff
@@ -343,3 +343,63 @@ exports.incProjectProgress = functions.database.ref('/v2/projects/{projectId}/re
...
+    try {
+        const userRef = await admin.database().ref('v2/users').once('value');
+        const data = userRef.val();
+
+        if (data) {
+            const newUserData: {[key: string]: string} = {};
+
+            Object.keys(data).forEach((id) => {
+                if (data[id]?.username) {
+                    const newUserNameKey = data[id].username?.toLowerCase() as string;
```
Comment (tnagorra) [inline firebase/functions/src/index.ts:387]: Not sure if we have already removed whitespaces from username. We might need to think about it.

File: manager-dashboard/app/views/NewProject/index.tsx [hunk 4]
```diff
@@ -271,9 +274,69 @@ function NewProject(props: Props) {
...
+
+                const uploadData = {
+                    ...finalValues,
+                    projectTopicKey,
+                    createdAt: (new Date()).getTime(),
+                };
+
+                const putProjectRef = databaseRef(db, `v2/projects/${newKey}`);
+                await setToDatabase(putProjectRef, uploadData);
```
Comment (tnagorra) [inline manager-dashboard/app/views/NewProject/index.tsx:327]: What is the difference between v2/projects and v2/projectDrafts? We are also creating projectDrafts later

**Review by tnagorra — APPROVED** [review]

**Review by thenav56 — APPROVED** [review]

**Pull request merged**